x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java

```diff
@@ -236,11 +247,28 @@ private List<ShardSearchFailure> selectFailures() {
     private void sendOneNodeRequest(TargetShards targetShards, ComputeListener computeListener, NodeRequest request) {
         final ActionListener<List<DriverProfile>> listener = computeListener.acquireCompute();
         sendRequest(request.node, request.shardIds, request.aliasFilters, new NodeListener() {
+
+            private final Set<ShardId> pendingRetries = new HashSet<>();
+
             void onAfter(List<DriverProfile> profiles) {
                 nodePermits.get(request.node).release();
                 if (concurrentRequests != null) {
                     concurrentRequests.release();
                 }
+
+                if (pendingRetries.isEmpty() == false && remainingTargetShardSearchAttempts.decrementAndGet() > 0) {
+                    ongoingTargetShardResolutionAttempts.incrementAndGet();
+                    var indices = pendingRetries.stream().map(ShardId::getIndexName).distinct().toArray(String[]::new);
+                    searchShards(indices, pendingRetries::contains, computeListener.acquireAvoid().delegateFailure((l, newSearchShards) -> {
```

Comment:
I think we need to make changes to the search_shards API to allow bypassing can_match in this case. We only need the up-to-date routing table here.

Reply:
Good point. I think we can have a separate action for it to make it simpler.